File: services/chimera/src/commands/clone.rs
use super::get_base64_image_data;
use async_trait::async_trait;
use duvua_cache::{redis::RedisCacheService, utils::get_or_store_user};
use duvua_framework::{
    builder::interaction_response::InteractionResponse,
    errors::BotError,
    handler::{CommandHandler, CommandHandlerData, CustomCommandType},
    utils::{get_avatar_url, get_option},
};
use serde_json::json;
use serenity::{
    builder::{CreateApplicationCommand, CreateApplicationCommandOption},
    model::prelude::{
        application_command::ApplicationCommandInteraction, command::CommandOptionType,
    },
    prelude::Context,
};
use std::{sync::Arc, time::Duration};

pub struct CloneCommand {
    data: &'static CommandHandlerData,
    cache: Arc<RedisCacheService>,
}

impl CloneCommand {
    pub fn new(cache: Arc<RedisCacheService>) -> Self {
        Self {
            data: Box::leak(Box::new(build_data())),
            cache,
        }
    }
}

#[async_trait]
impl CommandHandler for CloneCommand {
    async fn handle_command(
        &self,
        ctx: &Context,
        interaction: &ApplicationCommandInteraction,
    ) -> Result<(), BotError> {
        let user_id: u64 = get_option(&interaction.data.options, "user")
            .ok_or(BotError::OptionNotProvided("user"))?
            .value
            .ok_or(BotError::InvalidOption("user"))?
            .as_str()
            .ok_or(BotError::InvalidOption("user"))?
            .parse()
            .or(Err(BotError::InvalidOption("user")))?;

        let message = get_option(&interaction.data.options, "message")
            .ok_or(BotError::OptionNotProvided("message"))?
            .value
            .ok_or(BotError::InvalidOption("message"))?
            .as_str()
            .ok_or(BotError::InvalidOption("message"))?
            .to_owned();

        let user = get_or_store_user(&ctx.http, &self.cache, user_id).await?;

        let avatar = user.avatar.ok_or(BotError::UserAvatarFetchFailed)?;
        let avatar = get_avatar_url(user_id, &avatar, "webp", Some(128));

        let avatar_base64 = match get_base64_image_data(&avatar).await {
            Ok(v) => Some("data:image/webp;base64,".to_owned() + &v),
            Err(_) => None,
        };

        let data = json!({
            "name": user.name,
            "avatar": avatar_base64,
        });

        let webhook = ctx
            .http
            .create_webhook(interaction.channel_id.0, &data, Some("Clone command"))
            .await
            .or_else(|e| Err(BotError::Serenity(e)))?;

        _ = InteractionResponse::with_content_ephemeral("Clone criado")
            .respond(&ctx.http, interaction.id.0, &interaction.token)
            .await;

        if let Err(e) = webhook
            .execute(&ctx.http, false, |w| w.content(message))
            .await
        {
            log::error!("Failed to send webhook message: {e}");
        }

        tokio::time::sleep(Duration::from_millis(3000)).await;

        if let Err(e) = webhook.delete(&ctx.http).await {
            log::error!("Failed to delete webhook: {e}");
        }

        Ok(())
    }

    fn get_data(&self) -> &'static CommandHandlerData {
        self.data
    }
}

#[inline]
fn build_data() -> CommandHandlerData {
    CommandHandlerData {
        needs_defer: false,
        command_data: Some(build_data_command()),
        custom_id: None,
        kind: CustomCommandType::Fun,
    }
}

fn build_data_command() -> CreateApplicationCommand {
    CreateApplicationCommand::default()
        .name("clone")
        .description("Clone um usuário e faça o clone enviar uma mensagem")
        .description_localized("en-US", "Create a user clone and make it send some message")
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::User)
                .name("user")
                .description("O usuário que deseja clonar")
                .description_localized("en-US", "The user you want to clone")
                .required(true)
                .to_owned(),
        )
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::String)
                .name("message")
                .description("A mensagem que o clone irá mandar")
                .description_localized("en-US", "The message's the clone will send")
                .required(true)
                .to_owned(),
        )
        .to_owned()
}

File: internal/commands/clone.go
package commands

import (
	"bytes"
	"encoding/base64"
	"io"
	"log/slog"
	"mime"
	"net/http"

	"github.com/bwmarrin/discordgo"
	"github.com/zanz1n/duvua-bot/internal/errors"
	"github.com/zanz1n/duvua-bot/internal/manager"
	"github.com/zanz1n/duvua-bot/internal/utils"
)

var cloneCommandData = discordgo.ApplicationCommand{
	Name:        "clone",
	Description: "Clone um usuário e faça o clone enviar uma mensagem",
	DescriptionLocalizations: &map[discordgo.Locale]string{
		discordgo.EnglishUS: "Create a user clone and make it send some message",
	},
	Options: []*discordgo.ApplicationCommandOption{
		{
			Type:        discordgo.ApplicationCommandOptionUser,
			Name:        "user",
			Description: "O usuário que deseja clonar",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "The user you want to clone",
			},
			Required: true,
		},
		{
			Type:        discordgo.ApplicationCommandOptionString,
			Name:        "message",
			Description: "A mensagem que o clone irá mandar",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "The message's the clone will send",
			},
			Required: true,
		},
	},
}

func NewCloneCommand() manager.Command {
	return manager.Command{
		Accepts: manager.CommandAccept{
			Slash:  true,
			Button: false,
		},
		Data:       &cloneCommandData,
		Category:   manager.CommandCategoryFun,
		NeedsDefer: false,
		Handler:    &CloneCommand{},
	}
}

type CloneCommand struct {
}

func (c *CloneCommand) getBase64Avatar(hs *http.Client, guildId string, member *discordgo.Member) (string, error) {
	var url string
	if member.Avatar == "" {
		url = discordgo.EndpointUserAvatar(member.User.ID, member.User.Avatar)
	} else {
		url = discordgo.EndpointGuildMemberAvatar(guildId, member.User.ID, member.Avatar)
	}

	res, err := hs.Get(url)
	if err != nil {
		return "", err
	}

	if res.StatusCode != http.StatusOK {
		return "", errors.Unexpected("unexpected status code")
	} else if mt, _, _ := mime.ParseMediaType(res.Header.Get("Content-Type")); mt != "image/png" {
		return "", errors.Unexpectedf("unexpected image mime type `%s`", mt)
	}

	buf := bytes.NewBuffer([]byte{})

	_, err = io.Copy(buf, res.Body)
	if err != nil {
		return "", err
	}

	base64d := base64.StdEncoding.EncodeToString(buf.Bytes())

	return "data:image/png;base64," + base64d, nil
}

func (c *CloneCommand) Handle(s *discordgo.Session, i *discordgo.InteractionCreate) error {
	if i.Member == nil {
		return errors.New("esse comando só pode ser utilizado dentro de um servidor")
	}

	if i.Type != discordgo.InteractionApplicationCommand &&
		i.Type != discordgo.InteractionApplicationCommandAutocomplete {
		return errors.New("interação de tipo inesperado")
	}

	data := i.ApplicationCommandData()

	userOpt := utils.GetOption(data.Options, "user")
	if userOpt == nil {
		return errors.New("opção `user` é necessária")
	} else if userOpt.Type != discordgo.ApplicationCommandOptionUser {
		return errors.New("opção `user` precisa ser um usuário válido")
	}
	userId := userOpt.UserValue(nil).ID

	messageOpt := utils.GetOption(data.Options, "message")
	if messageOpt == nil {
		return errors.New("opção `message` é necessária")
	} else if messageOpt.Type != discordgo.ApplicationCommandOptionString {
		return errors.New("opção `user` precisa ser um texto")
	}
	message := messageOpt.StringValue()

	member, err := s.State.Member(i.GuildID, userId)
	if err != nil {
		member, err = s.GuildMember(i.GuildID, userId)
		if err != nil {
			return err
		}
	}

	avatar, err := c.getBase64Avatar(s.Client, i.GuildID, member)
	if err != nil {
		slog.Error(
			"Failed to fetch user avatar on discord cdn",
			"user_id", userId,
			"error", err,
		)
	}

	auditR := discordgo.WithAuditLogReason("Comando /clone")

	webhook, err := s.WebhookCreate(i.ChannelID, member.DisplayName(), avatar, auditR)
	if err != nil {
		slog.Error("Failed to create discord webhook", "error", err)
		return errors.New("não foi possível criar o clone")
	}

	defer func() {
		if err := s.WebhookDelete(webhook.ID, auditR); err != nil {
			slog.Error("Failed to delete discord webhook", "error", err)
		}
	}()

	params := discordgo.WebhookParams{Content: message}

	_, err = s.WebhookExecute(webhook.ID, webhook.Token, false, &params)
	if err != nil {
		slog.Error("Failed to send webhook message", "webhook_id", webhook.ID, "error", err)
		return errors.New("não foi possível enviar a mensagem desejada")
	}

	return s.InteractionRespond(i.Interaction, utils.BasicResponse("Clone criado!"))
}
